### DISCUSSION TITLE
How can I instantiate the main Slint component for tests multiple times?

### DISCUSSION THREAD

**Discussion by alexeyabel** [slint-ui/slint#6383] [Q&A]
I am looking for a way to instantiate the main component multiple times to execute tests.

I have a model struct `Model` that takes a weak reference to the `MainWindow` Slint component as a constructor argument like this

```rust
pub struct Model {
    ui_weak: Weak<MainWindow>,
    settings: Settings,
    // ...
}

impl Model {
    pub fn new(settings: Settings, ui_weak: Weak<MainWindow>) -> Model {
        Model {
            ui_weak,
            settings: settings.clone(),
            // ...
        }
    }
```

Later in tests I have 

```rust
#[cfg(test)]
mod tests {
    use super::*;
    //use serial_test::serial; // <-- added in second step
    use slint::ComponentHandle;

    #[test]
    //#[serial] // <-- added in second step
    fn my_first_test() {
        let ui_weak = MainWindow::new().unwrap().as_weak();
        let settings = Settings::default();
        let model = Model::new(settings, ui_weak);
        // ...
    }

    #[test]
    //#[serial] // <-- added in second step
    fn my_second_test() {
        let ui_weak = MainWindow::new().unwrap().as_weak();
        let settings = Settings::default();
        let model = Model::new(settings, ui_weak);
        // ...
    }
```

This fails with
```
WARNING: QApplication was not created in the main() thread.
```
I tried using the serial_test crate and added `#[serial]` before each test, but this time I got:

```
---- model::tests::my_first_test stdout ----
thread 'model::tests::my_first_test' panicked at src/model.rs:274:41:
called `Result::unwrap()` on an `Err` value: SetPlatformError(AlreadySet)
```

How can I instantiate the main component for tests? I am on Linux, NixOS, KDE, Qt5.

**Comment by ogoffart** [comment]
The problem is that the platform (Qt or winit) won't allow that.

But there is a way: develop your own platform for the test:
See what we are doing there.
https://github.com/slint-ui/slint/blob/master/internal/backends/testing/testing_backend.rs

But there is still a catch that this backend had two entry points: (define in testing/lib.rs)
 - `init_no_event_loop` this create a backend that can be instantiated from thread but cannot use `invoke_from_event_loop` or `spawn_locale`
 - `init_integration_test_with_*` can use everything but cannot be used in a thread.

I think the best option right now is to do "integration" test in their own files so they own get their own process. instead of different thread.
